sympy/core/function.py

```diff
@@ -1213,6 +1213,10 @@ def __new__(cls, expr, *variables, **kwargs):
         # derivative class by calling Expr.__new__.
         if (not (hasattr(expr, '_eval_derivative') and evaluate) and
            (not isinstance(expr, Derivative))):
+            # If we try to differentiate a variable with respect to itself,
+            # the result should always be 1.
```

Comment:
Well, _always_ is not correct. Think about matrices.

Reply:
While it's true that, in a general mathematical basis, there are concepts called differentiation for which it is not the case that `d/dx (x) == 1`, it is consistent with SymPy's usage. For example, `diff(2 * x, x) == 2`, even if `x` is completely unknown — which, likewise, is not valid for matrices. On top of this, SymPy can't (currently) even differentiate matrices with respect to other matrices, so this particular example probably isn't going to cause any issues directly.
I agree in general, but SymPy makes enough assumptions here that I think it's fair to interpret the comment in the context of the `diff` function.

Reply:
> On top of this, SymPy can't (currently) even differentiate matrices with respect to other matrices, so this particular example probably isn't going to cause any issues directly.

This is something we are trying to implement, though. 

Matrices are going to have to define `_eval_derivative` anyway. So the question is should we also require every special `_diff_wrt` object to define this as well, or should we implement the common case here so that it isn't required for most things (at least, anything whose derivative is a scalar).

Reply:
> So the question is should we also require every `special _diff_wrt` object to define this as well, or should we implement the common case here so that it isn't required for most things (at least, anything whose derivative is a scalar).

Yeah — this comes down to a decision decision: treating values as scalars by default or not. Reasonably either this case should be handled, or all the current scalar assumptions should be removed.